src/plonk/permutation/prover.rs

```diff
@@ -166,76 +217,125 @@ impl<C: CurveAffine> Committed<C> {
-            // l_0(X) * (1 - z(X)) = 0
-            .chain(Some(
-                Polynomial::one_minus(self.permutation_product_coset.clone()) * &pk.l0,
-            ))
-            // z(X) \prod (p(X) + \beta s_i(X) + \gamma) - z(omega^{-1} X) \prod (p(X) + \delta^i \beta X + \gamma)
-            .chain(Some({
-                let mut left = self.permutation_product_coset.clone();
-                for (values, permutation) in p
-                    .columns
+            // Enforce only for the first set.
+            // l_0(X) * (1 - z_0(X)) = 0
+            .chain(self.sets.first().map(|first_set| {
+                Polynomial::one_minus(first_set.permutation_product_coset.clone()) * &pk.l0
+            }))
+            // Enforce only for the last set.
+            // l_last(X) * (z_l(X)^2 - z_l(X)) = 0
+            .chain(self.sets.last().map(|last_set| {
+                ((last_set.permutation_product_coset.clone() * &last_set.permutation_product_coset)
+                    - &last_set.permutation_product_coset)
+                    * &pk.l_last
+            }))
+            // Except for the first set, enforce.
+            // l_0(X) * (z_i(X) - z_{i-1}(\omega^(last) X)) = 0
+            .chain(
+                self.sets
                     .iter()
-                    .map(|&column| match column.column_type() {
-                        Any::Advice => {
-                            &advice_cosets[pk.vk.cs.get_any_query_index(column, Rotation::cur())]
-                        }
-                        Any::Fixed => {
-                            &fixed_cosets[pk.vk.cs.get_any_query_index(column, Rotation::cur())]
-                        }
-                        Any::Instance => {
-                            &instance_cosets[pk.vk.cs.get_any_query_index(column, Rotation::cur())]
-                        }
+                    .skip(1)
+                    .zip(self.sets.iter())
+                    .map(|(set, last_set)| {
+                        (
+                            set.permutation_product_coset.clone(),
+                            last_set.permutation_product_coset_last.clone().unwrap(),
+                        )
                     })
-                    .zip(pkey.cosets.iter())
-                {
-                    parallelize(&mut left, |left, start| {
-                        for ((left, value), permutation) in left
-                            .iter_mut()
-                            .zip(values[start..].iter())
-                            .zip(permutation[start..].iter())
+                    .collect::<Vec<_>>()
+                    .into_iter()
+                    .map(move |(coset, coset_last)| (coset - &coset_last) * &pk.l0),
+            )
+            // And for all the sets we enforce:
+            // (1 - (l_last + l_blind)) * (
+            //   z_i(\omega X) \prod (p(X) + \beta s_i(X) + \gamma)
+            // - z_i(X) \prod (p(X) + \delta^i \beta X + \gamma)
```

Comment:
\prod over what range? (See #320 for what I think it is.)

Reply:
I meant for this to be an informal/implicit product but I did omit the correct indicies, I'll fix that to clarify.